lib/sliceutil/sliceutil.go

```diff
@@ -0,0 +1,16 @@
+// Copyright (C) 2023 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package sliceutil
+
+// RemoveAndZero removes the element at index i from slice s and returns the
+// resulting slice. The slice ordering is preserved; the last slice element
+// is zeroed before shrinking.
+func RemoveAndZero[E any, S ~[]E](s S, i int) S {
+	copy(s[i:], s[i+1:])
+	s[len(s)-1] = *new(E)
+	return s[:len(s)-1]
+}
```

Comment:
I am missing something about the zero-ing: We are zeroing the element at the end, but that element is still referenced at the second to last (resp. new last in the truncated slice) - as in it can't be GCed. Why are we zeroing then?

Reply:
This is the typical slice removal pattern. :) It's just about making sure elements no longer part of the slice (but still present in the backing array) do not contain pointers pointing to things that prevent them from being GC:d.

Reply:
I get the general mechanics, it's just the zeroing that confused me - I think I get it now. I think the zeroing only has an effect though if we are removing the last element - otherwise the element in the backing array after the slice end is also present in the slice due to the copy - so zeroing it out doesn't do anything for GC. Only if it happens to be the last element that element would persist in the array but not the slice. Hardly worth "optimising" (moving the zeroing into a `if i == len(s) - 1` block) though, I am just happy I didn't have a fundamental misunderstanding here (hopefully) :)

Reply:
> otherwise the element in the backing array after the slice end is also present in the slice due to the copy - so zeroing it out doesn't do anything for GC

Not this cycle, no, but then you've set it up for *that* element never being GC:d.